unison-share-api/src/Unison/Server/Share/FuzzyFind.hs

```diff
@@ -0,0 +1,364 @@
+import Unison.Codebase qualified as Codebase
+import Unison.Codebase.Editor.DisplayObject
+import Unison.Codebase.Path qualified as Path
+import Unison.Codebase.ShortCausalHash qualified as SCH
+import Unison.Codebase.SqliteCodebase.Conversions qualified as Cv
+import Unison.NameSegment
+import Unison.Parser.Ann (Ann)
+import Unison.Prelude
+import Unison.PrettyPrintEnvDecl qualified as PPED
+import Unison.PrettyPrintEnvDecl.Sqlite qualified as PPED
+import Unison.Server.Backend (termEntryLabeledDependencies, typeEntryLabeledDependencies)
+import Unison.Server.Backend qualified as Backend
+import Unison.Server.Syntax (SyntaxText)
+import Unison.Server.Types
+  ( APIGet,
+    ExactName (..),
+    HashQualifiedName,
+    NamedTerm,
+    NamedType,
+    UnisonName,
+    mayDefaultWidth,
+  )
+import Unison.Symbol (Symbol)
+import Unison.Util.Pretty (Width)
+
+type FuzzyFindAPI =
+  "find"
+    :> QueryParam "rootBranch" SCH.ShortCausalHash
+    :> QueryParam "relativeTo" Path.Path
+    :> QueryParam "limit" Int
+    :> QueryParam "renderWidth" Width
+    :> QueryParam "query" String
+    :> APIGet [(Alignment, FoundResult)]
+
+instance ToSample Alignment where
+  toSamples _ = noSamples
+
+instance ToParam (QueryParam "limit" Int) where
+  toParam _ =
+    DocQueryParam
+      "limit"
+      ["1", "10", "20"]
+      "The maximum number of results to return. Defaults to 10."
+      Normal
+
+instance ToParam (QueryParam "query" String) where
+  toParam _ =
+    DocQueryParam
+      "query"
+      ["foo", "ff", "td nr"]
+      "Space-separated subsequences to find in the name of a type or term."
+      Normal
+
+data FoundTerm = FoundTerm
+  { bestFoundTermName :: HashQualifiedName,
+    namedTerm :: NamedTerm
+  }
+  deriving (Generic, Show)
+
+data FoundType = FoundType
+  { bestFoundTypeName :: HashQualifiedName,
+    typeDef :: DisplayObject SyntaxText SyntaxText,
+    namedType :: NamedType
+  }
+  deriving (Generic, Show)
+
+instance ToJSON FoundType where
+  toJSON (FoundType {bestFoundTypeName, typeDef, namedType}) =
+    object
+      [ "bestFoundTypeName" .= bestFoundTypeName,
+        "typeDef" .= typeDef,
+        "namedType" .= namedType
+      ]
+
+deriving instance ToSchema FoundType
+
+instance ToJSON FoundTerm where
+  toJSON (FoundTerm {bestFoundTermName, namedTerm}) =
+    object
+      [ "bestFoundTermName" .= bestFoundTermName,
+        "namedTerm" .= namedTerm
+      ]
+
+deriving instance ToSchema FoundTerm
+
+data FoundResult
+  = FoundTermResult FoundTerm
+  | FoundTypeResult FoundType
+  deriving (Generic, Show)
+
+instance ToJSON FoundResult where
+  toJSON = \case
+    FoundTermResult ft -> object ["tag" .= String "FoundTermResult", "contents" .= ft]
+    FoundTypeResult ft -> object ["tag" .= String "FoundTypeResult", "contents" .= ft]
+
+deriving instance ToSchema FoundResult
+
+instance ToSample FoundResult where
+  toSamples _ = noSamples
+
+serveFuzzyFind ::
+  -- | Whether the root is a scratch root
+  Bool ->
```

Comment:
@ChrisPenner Because this was a new module, I missed that this was a new parameter. Can you document a bit more about what the effect of this parameter is?

Should this module even be in `unison` or should it be in `enlil`?

Reply:
e.g. is UCM calling this API?

Reply:
UCM doesn't ever call this API.

This param is there so we can support the "Allow searching all your projects from the root of a scratch project" like we discussed.

I'll probably eventually move some of this to enlil, but I've been delaying it as long as possible because keeping it in unison means that the code stays up to date with refactors, otherwise I worry that large sweeping refactors in unison will leave me a ton of work in enlil that got missed because it didn't get caught in `unison`'s CI.

Reply:
Yeah I only spotted it while reviewing the Enlil pr. Ok so the idea is that it's a flag that you set to true when `rootCausal` is a "scratch" root, and then the function will do whatever-we-decided-was-appropriate-for-scratch-root. Is that right?

re sweeping refactors, yeah, we need to change something to reduce the amount of work there; maybe:
* break up modules differently to have less coupling?
* have a CI test for `unison` that checks a compatible enlil build (tricky, but that's essentially what's been proposed int he past)
* other?